Reduce memory usage when scheduler start jobs

nickva opened the pull request:
Scheduler calls pending_jobs which returns all available jobs. This could be
a huge list of all job records. Instead use a smaller list of tuples, sort
and filter those then start jobs based on job ids.

sagelywizard commented on src/couch_replicator_scheduler.erl:
```diff
@@ -381,9 +387,19 @@ pending_job_count() ->
+% A smaller tuple is used here to reduce memory usage due to returning
+% potentially a huge list of pending job records.
+-spec pending_jobs_info() -> [{job_id(), erlang:timestamp()}].
+pending_jobs_info() ->
```
What do you think about passing in `Count` to `pending_jobs_info` and only accumulating the top `Count` entries? This still accumulates an element for ~every entry in the table.

nickva replied:
@sagelywizard Great idea. let's try that. Perhaps have a sorted set (`gb_sets`). If its size has grown above the `Count` and current element is older than youngest one. Add element in set and remove youngest. Take will take more CPU time but code might look cleaner and memory should look better as well.

nickva committed:
Reduce memory usage when scheduler starts jobs

Scheduler's `pending_jobs/0` returned all available jobs. This could be
a huge list of job records. The list was then sorted and only a few jobs
were picked from the front (usually as a default). This result in large
amounts of garbage (on the order of 0.5G for 33k replication jobs).

Instead, use an ets:foldl with a sorted set accumulator to gather pending jobs.

nickva commented:
@rnewson @sagelywizard Hot-patched this code a test cluster. So far see memory size reduction from 800MB to 1MB.

sagelywizard commented:
You know, if this change as-is reduces the memory footprint by 100x, it might not be worth adding the further gb_trees-type optimization which you mentioned. Just a thought. Can talk about it on Slack tomorrow.

rnewson commented on src/couch_replicator_scheduler.erl:
```diff
@@ -271,12 +271,54 @@ handle_config_terminate(Self, _, _) ->
+pending_jobs(Count) when is_integer(Count), Count > 0 ->
+    Set0 = gb_sets:new(),  % [{LastStart, Job},...]
+    Now = os:timestamp(),
+    {Set1, _, _} = ets:foldl(fun pending_fold/2, {Set0, Now, Count}, ?MODULE),
```
a record instead of the 3-tuple would be more readable.

rnewson replied:
discuss.

rnewson commented on src/couch_replicator_scheduler.erl:
```diff
@@ -271,12 +271,54 @@ handle_config_terminate(Self, _, _) ->
+             % have to replace one of the accumulated jobs
+             pending_maybe_replace(Job, Set);
+        {true, false} ->
+             % Jobs is healthy and but haven't reached the limit, so add job
```
"Job"

rnewson replied:
s/and but/and we/

rnewson commented:
+1
